### ISSUE TITLE
barman recover doesn't allow cross-timeline restore of WAL

### ISSUE THREAD

**Issue by gclough** [2ndquadrant-it/barman#178]
When doing a `barman recover` using a command such as this, you cannot automatically recover across multiple timelines:

```$ barman recover --get-wal --remote-ssh-command "ssh postgres@dbserver01" db01 20180731T200002 /var/lib/pgsql/9.6/data```

Which is restoring a backup taken several timelines ago:

```
$ barman list-backup db01
db01 20180814T200001 - Wed Aug 15 00:00:20 2018 - Size: 135.3 MiB - WAL Size: 80.0 MiB
db01 20180813T200001 - Tue Aug 14 00:00:13 2018 - Size: 135.1 MiB - WAL Size: 80.0 MiB
db01 20180812T200001 - Mon Aug 13 00:00:12 2018 - Size: 135.0 MiB - WAL Size: 64.0 MiB
db01 20180811T200001 - Sun Aug 12 00:00:19 2018 - Size: 135.2 MiB - WAL Size: 64.0 MiB
db01 20180810T200001 - Sat Aug 11 00:00:14 2018 - Size: 138.2 MiB - WAL Size: 64.0 MiB
db01 20180809T200002 - Fri Aug 10 00:00:17 2018 - Size: 142.0 MiB - WAL Size: 80.0 MiB
db01 20180808T200002 - Thu Aug  9 00:00:24 2018 - Size: 142.3 MiB - WAL Size: 80.0 MiB
db01 20180807T200001 - Wed Aug  8 00:00:26 2018 - Size: 142.1 MiB - WAL Size: 80.0 MiB
db01 20180806T200002 - Tue Aug  7 00:00:21 2018 - Size: 147.1 MiB - WAL Size: 80.0 MiB
db01 20180805T200001 - Mon Aug  6 00:00:34 2018 - Size: 147.1 MiB - WAL Size: 96.0 MiB
db01 20180804T200001 - Sun Aug  5 00:00:27 2018 - Size: 238.7 MiB - WAL Size: 96.0 MiB
db01 20180803T200001 - Sat Aug  4 00:00:04 2018 - Size: 238.8 MiB - WAL Size: 96.0 MiB
db01 20180802T200001 - Fri Aug  3 00:00:02 2018 - Size: 234.3 MiB - WAL Size: 96.0 MiB
db01 20180801T200001 - Thu Aug  2 00:00:03 2018 - Size: 220.0 MiB - WAL Size: 416.0 MiB
db01 20180731T200002 - Wed Aug  1 00:00:02 2018 - Size: 192.9 MiB - WAL Size: 0 B
db01 20180730T200002 - Tue Jul 31 00:00:02 2018 - Size: 192.9 MiB - WAL Size: 0 B
db01 20180729T200001 - Mon Jul 30 00:00:02 2018 - Size: 192.9 MiB - WAL Size: 0 B
db01 20180728T200002 - Sun Jul 29 00:00:02 2018 - Size: 192.9 MiB - WAL Size: 0 B
db01 20180727T200002 - Sat Jul 28 00:00:02 2018 - Size: 192.9 MiB - WAL Size: 0 B
db01 20180726T200001 - Fri Jul 27 00:00:02 2018 - Size: 192.9 MiB - WAL Size: 0 B
db01 20180725T200002 - Thu Jul 26 00:00:02 2018 - Size: 192.9 MiB - WAL Size: 0 B
db01 20180724T200002 - Wed Jul 25 00:00:02 2018 - Size: 192.9 MiB - WAL Size: 0 B
db01 20180723T200002 - Tue Jul 24 00:00:02 2018 - Size: 192.9 MiB - WAL Size: 0 B
db01 20180722T200001 - Mon Jul 23 00:00:02 2018 - Size: 192.9 MiB - WAL Size: 0 B
db01 20180721T200001 - Sun Jul 22 00:00:02 2018 - Size: 192.9 MiB - WAL Size: 0 B
db01 20180720T200002 - Sat Jul 21 00:00:02 2018 - Size: 192.9 MiB - WAL Size: 0 B
db01 20180719T200002 - Fri Jul 20 00:00:02 2018 - Size: 192.9 MiB - WAL Size: 0 B
db01 20180718T200002 - Thu Jul 19 00:00:02 2018 - Size: 192.9 MiB - WAL Size: 0 B
db01 20180717T200001 - Wed Jul 18 00:00:02 2018 - Size: 192.9 MiB - WAL Size: 0 B
```

The `recovery.conf` that's produced looks like this:

```
$ cat recovery.conf
# The 'barman-wal-restore' command is provided in the 'barman-cli' package
restore_command = 'barman-wal-restore -U barman dbserver01 db01 %f %p'
```

That stops recovery when it hits a timeline change, so complete recovery is not possible unless `recovery_target_timeline` is added:

```
$ cat recovery.conf
# The 'barman-wal-restore' command is provided in the 'barman-cli' package
restore_command = 'barman-wal-restore -U barman dbserver01 db01 %f %p'
recovery_target_timeline = 'latest'
```

**ACTION:**  Could you please either make this standard practice, or at least give an option to add it to the recovery.conf automatically?

**Comment by xilu0** [comment]
why the issue is opening for a long time, but nobody responds

**Comment by xilu0** [comment]
> why the issue is opening for a long time, but nobody responds

oh, it's already been in a milestone, good!

**Comment by xilu0** [comment]
I want to help developing barman, how can I do?

**Comment by mikewallace1979** [comment]
Hi @xilu0 - that's great, we'd love to have more developers involved in the project!

The current status for contributing is that anybody is welcome to submit a PR (there is some basic guidance in the docs at http://docs.pgbarman.org/release/2.13/#contributing-to-barman) and we'll consider it for merging.

**Comment by mikewallace1979** [comment]
Once https://github.com/EnterpriseDB/barman/issues/305 is implemented it will be possible to use `--target-tli=latest` or `--target-tli=current` to have barman set `recovery_target_timeline` to the respective value - this would meet the requirement of being able to have barman automatically set the correct value in the PostgreSQL recovery config.

However, I am undecided about making it a standard practice to set `recovery_target_timeline = 'latest'`. On PostgreSQL>=12 this is actually the default, so setting it would not be necessary. Because the original issue references `recovery.conf` we know that the issue was encountered using PostgreSQL<12, and the default value of `recovery_target_timeline` in <12 is [current](https://www.postgresql.org/docs/11/recovery-target-settings.html), which meansrecovery will stop when it gets to the most recent WAL on the same timeline as the backup being restored (I believe this matches the behaviour described in the issue).

So we could change Barman so that it sets `recovery_target_timeline = 'latest'` in `recovery.conf` if no `--target-tli` is specified. My inclination is *not* to do this because:
1. Barman would default to `latest` against PostgreSQL<12, where PostgreSQL<12 defaults to `current` - choosing different default behaviour to PostgreSQL is something we generally want to avoid.
2. The default behaviour of Barman would change when running against PostgreSQL<12. If we're going to do this we certainly shouldn't do it on a minor version release as it is a breaking change which could break recovery scenarios for anyone relying on defaulting to the timeline of the backup being restored in PostgreSQL<12.

In conclusion, I suspect implementing #305 and updating the documentation clarifying the need to use `--target-tli=latest` when running against PostgreSQL<12 would be enough to resolve this issue without introducing any breaking changes or conflicts between Barman and PostgreSQL defaults.

@amenonsen what do you think?

**Comment by amenonsen** [comment]
I agree. Being able to add `--target-tli=latest` is sufficient, let's not change any defaults.

**Comment by mikewallace1979** [comment]
> In conclusion, I suspect implementing https://github.com/EnterpriseDB/barman/issues/305 and updating the documentation clarifying the need to use --target-tli=latest when running against PostgreSQL<12 would be enough to resolve this issue without introducing any breaking changes or conflicts between Barman and PostgreSQL defaults.

Given #305 is now merged I think we are good to update the docs as discussed.
